### ISSUE TITLE
GeneratorConfiguration fails in Kotlin

### ISSUE THREAD

**Issue by vitekkor** [rohanpadhye/JQF#130]
Hi! I want to use your library for testing using zest. I wrote a test driver in java according to how you described it. Everything works great. Things get bad when I try to rewrite the test driver in Kotlin - the InRange annotation is not applied to the generator and it generates numbers in the range from Integer.MIN_VALUE to Integer.MAX_VALUE. I tried to debug this and it looks like jqf just doesn't see annotations from Kotlin.

Here is my java code
```java
@RunWith(JQF.class)
public class ZestTestJava {
    @Fuzz
    public void angleInRadianTest(
            @InRange(minInt = 0, maxInt = 359) @From(IntegerGenerator.class) int grad,
            @InRange(minInt = 0, maxInt = 59) @From(IntegerGenerator.class) int min,
            @InRange(minInt = 0, maxInt = 59) @From(IntegerGenerator.class) int sec
    ) {
        double model = ((grad * 60 + min) * 60 + sec) / 206264.806247;
        Assert.assertEquals(model, lesson1.task1.SimpleKt.angleInRadian(grad, min, sec), 1e-5);
    }
}
```

Kotlin
```kotlin
@RunWith(JQF::class)
class ZestTests {

    @Fuzz
    fun angleInRadianTest(
        grad: @InRange(minInt = 0, maxInt= 359) @From(IntegerGenerator::class) Int,
        min: @InRange(minInt = 0, maxInt = 59) @From(IntegerGenerator::class) Int,
        sec: @InRange(minInt = 0, maxInt = 59) @From(IntegerGenerator::class) Int
    ) {
        val model = ((grad * 60 + min) * 60 + sec) / 206264.806247
        Assert.assertEquals(model, angleInRadian(grad, min, sec), 1e-5)
    }
}
```

**Comment by rohanpadhye** [comment]
Thanks for the issue report! This is interesting. The range annotations etc. are from [junit-quickcheck](https://github.com/pholser/junit-quickcheck), so let's try to figure out if this is something I can fix in JQF. Can you please try your example by replacing the `@RunWith` and `@Fuzz` annotations with the basic junit-quickcheck ones? 

```kotlin
@RunWith(JUnitQuickcheck::class)
class PropertyTests {

    @Property
    fun angleInRadianTest(
        grad: @InRange(minInt = 0, maxInt= 359) @From(IntegerGenerator::class) Int,
        min: @InRange(minInt = 0, maxInt = 59) @From(IntegerGenerator::class) Int,
        sec: @InRange(minInt = 0, maxInt = 59) @From(IntegerGenerator::class) Int
    ) {
        val model = ((grad * 60 + min) * 60 + sec) / 206264.806247
        Assert.assertEquals(model, angleInRadian(grad, min, sec), 1e-5)
    }
}
```

If you still see the problem, I'd be happy to file an issue upstream with [junit-quickcheck](https://github.com/pholser/junit-quickcheck) on this. If it works in this mode but not with JQF, then I'll have try to debug more---I have never used JQF with Kotlin before, but this is certainly a use case that I'd like to support if it's straightforward to do so.

**Comment by rohanpadhye** [comment]
Okay I tried and it seems like the problem persists in junit-quickcheck. There also seems to be an issue pholser/junit-quickcheck#77 that seems to have attempted a fix in pholser/junit-quickcheck@8a7bfa05067, but it doesn't seem like the test case above seems to get the range annotation.

**Comment by vitekkor** [comment]
I tried the code you wrote but nothing changed. 

>  ```kotlin
>
>@RunWith(JUnitQuickcheck::class)
>class PropertyTests {
>
>    @Property
>    fun angleInRadianTest(
>        grad: @InRange(minInt = 0, maxInt= 359) @From(IntegerGenerator::class) Int,
>        min: @InRange(minInt = 0, maxInt = 59) @From(IntegerGenerator::class) Int,
>        sec: @InRange(minInt = 0, maxInt = 59) @From(IntegerGenerator::class) Int
>    ) {
>        val model = ((grad * 60 + min) * 60 + sec) / 206264.806247
>        Assert.assertEquals(model, angleInRadian(grad, min, sec), 1e-5)
>    }
>}
>```

However, I decided to check one more option and wrote the following code. 
```kotlin
@RunWith(JUnitQuickcheck::class)
class ZestTests {

    @Property
    fun angleInRadianTest(
        @InRange(minInt = 0, maxInt = 59) @From(IntegerGenerator::class) grad: Int,
        @InRange(minInt = 0, maxInt = 59) @From(IntegerGenerator::class) min: Int,
        @InRange(minInt = 0, maxInt = 59) @From(IntegerGenerator::class) sec: Int
    ) {
        val model = ((grad * 60 + min) * 60 + sec) / 206264.806247
        Assert.assertEquals(model, angleInRadian(grad, min, sec), 1e-5)
    }
}
```
Surprisingly, the InRange annotation works. I tried posting annotations correctly and run with JQF.
```kotlin
@RunWith(JQF::class)
class ZestTests {

    @Fuzz
    fun angleInRadianTest(
        @InRange(minInt = 0, maxInt = 59) @From(IntegerGenerator::class) grad: Int,
        @InRange(minInt = 0, maxInt = 59) @From(IntegerGenerator::class) min: Int,
        @InRange(minInt = 0, maxInt = 59) @From(IntegerGenerator::class) sec: Int
    ) {
        val model = ((grad * 60 + min) * 60 + sec) / 206264.806247
        Assert.assertEquals(model, angleInRadian(grad, min, sec), 1e-5)
    }
}
```
and... It doesn't work.
It looks like the problem is in JQF

**Comment by rohanpadhye** [comment]
OK, pushed an update that should hopefully resolve this. Could you try with the latest version? If using the Maven plugin it would mean cloning the latest JQF repo, running `mvn install`, and depending on `jqf-fuzz` version `1.7-SNAPSHOT`.

If this works I'll package this in a 1.7 release along with several other pending updates.

**Comment by vitekkor** [comment]
Yes! It works. Thanks!

**Comment by vitekkor** [comment]
Yes! It works. But I noticed an interesting feature - the maxInt value in the InRange annotation is now excluded, that is, if you write `@InRange(minInt=0, maxInt=1)`, then always only the value 0 will be generated. At the same time, in the junit-quickcheck, it is generated randomly then 0, then 1.

**Comment by vitekkor** [comment]
It looks like the problem due to which when the range is 1 (InRange (minInt = 0, maxInt = 1)) only zero is generated, lies in this method.
[FastSourceOfRandomness](https://github.com/rohanpadhye/JQF/blob/master/fuzz/src/main/java/edu/berkeley/cs/jqf/fuzz/junit/quickcheck/FastSourceOfRandomness.java)
```java
    private int fastChooseIntInRange(int min, int max) {
        int range = max - min;

        // If range is too wide, overflow will make it negative
        if (range > 0) {
            int random = delegate.nextInt() % range;
            if (random < 0) {
                random += range;
            }
            return min + random;
        } else {
            return (int) Ranges.choose(this, min, max);
        }
    }
````
I tried other ranges (InRange (minInt = 40, maxInt = 41) for example) and there also only the lower bound is always generated, i.e. 40. That is, the upper bound is not included in the range.

It seems that if you rewrite the code like this, then everything works correctly.
```java
int random = this.delegate.nextInt() % (range + 1);
```
Similarly, you need to make changes to the `fastChooseByteInRange(byte min, byte max)` method.

**Comment by rohanpadhye** [comment]
Wow that's very interesting. The Java Random.nextInt(bound) method
considers the bound to be an exclusive upper bound of the range (consistent
with most random APIs), but it seems like junit-quickcheck treats the range
as inclusive?

I'll look into this further because simply making this change might break
existing JQF drivers if they rely on the bound behavior. But I agree that
we must keep consistency with quickcheck.

Thanks for sharing the observation and also for localizing the fault!

**Comment by vitekkor** [comment]
However, if memory serves me, in version 1.6 the upper limit was included in the range

**Comment by rohanpadhye** [comment]
The code you pointed to hasn't changed since 1.6 as far as I can tell, so
that can't be it.

**Comment by vitekkor** [comment]
Yes, that means it seemed to me. In `junit-quickchek`, the upper limit is inclusive, in `JQF` it is not. But I don't think this is a problem.

Overall my problem was resolved. Now I'm waiting for version 1.7. Thanks for the help!

**Comment by rohanpadhye** [comment]
Sounds good. Thanks for the detailed reports.

**Comment by rohanpadhye** [comment]
The updates are now available in JQF release 1.7. Thanks for the issue report!
